xds/internal/balancer/outlierdetection/e2e_test/outlierdetection_test.go

```diff
@@ -367,3 +383,69 @@ func (s) TestNoopConfiguration(t *testing.T) {
 		t.Fatalf("error in expected round robin: %v", err)
 	}
 }
+
+// Test verifies that outlier detection doesn't eject subchannels created by
+// the new pickfirst balancer when pickfirst is a non-leaf policy, i.e. not
+// under a petiole policy. When pickfirst is not under a petiole policy, it will
+// not register a health listener. pickfirst will still set the address
+// attribute to disable ejection through the raw connectivity listener. When
+// Outlier Detection processes a health update and sees the health listener is
+// enabled but a health listener is not registered, it will drop the ejection
+// update.
+func (s) TestPickFirstHealthListenerDisabled(t *testing.T) {
+	backend := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			return nil, errors.New("some error")
+		},
+	}
+	if err := backend.StartServer(); err != nil {
+		t.Fatalf("Failed to start backend: %v", err)
+	}
+	defer backend.Stop()
+	t.Logf("Started bad TestService backend at: %q", backend.Address)
+
+	countingODServiceConfigJSON := fmt.Sprintf(`
+	{
+	  "loadBalancingConfig": [
+		{
+		  "outlier_detection_experimental": {
+			"interval": "0.025s",
+			"baseEjectionTime": "0.100s",
+			"maxEjectionTime": "300s",
+			"failurePercentageEjection": {
+				"threshold": 50,
+				"enforcementPercentage": 100,
+				"minimumHosts": 0,
+				"requestVolume": 2
+			},
+			"childPolicy": [{"%s": {}}]
+		  }
+		}
+	  ]
+	}`, pickfirstleaf.Name)
+	sc := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(countingODServiceConfigJSON)
+
+	mr := manual.NewBuilderWithScheme("od-e2e")
+
+	mr.InitialState(resolver.State{
+		Addresses:     []resolver.Address{{Addr: backend.Address}},
+		ServiceConfig: sc,
+	})
+	cc, err := grpc.NewClient(mr.Scheme()+":///", grpc.WithResolvers(mr), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("grpc.NewClient() failed: %v", err)
+	}
+	defer cc.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	testServiceClient := testgrpc.NewTestServiceClient(cc)
+
+	// Failing request should not cause ejection.
+	testServiceClient.EmptyCall(ctx, &testpb.Empty{})
+	testServiceClient.EmptyCall(ctx, &testpb.Empty{})
+	testServiceClient.EmptyCall(ctx, &testpb.Empty{})
+	// Wait for the failure rate algorithm to run once.
+	shortCtx, shortCancel := context.WithTimeout(ctx, 50*time.Millisecond)
```

Comment:
IIUC, the ejection rate is once every 25ms?

Note that tests that are expecting "no change in behavior" before/after an operation are inherently flaky in their ability to detect the problem, since they are timing sensitive.  Is there any other way to test the same behaviors?

Reply:
> IIUC, the ejection rate is once every 25ms?

Yes, and the subchannel remains ejected for 100ms. So I was waiting to 2 * ejection rate to avoid flakiness.

> Is there any other way to test the same behaviors?

Yes, by moving this test to the od package, I can trigger the ejection algorithm to run directly instead of waiting for the passage of time. I made this change.